File: AocCsharp/Day03/Day03.cs
using AocCsharp.Utils;

namespace AocCsharp.Day03;

public static class Day03
{
    public static void Run()
    {
        var testInputPath = Path.Combine("..", "inputs", "day03_test.txt");
        var realInputPath = Path.Combine("..", "inputs", "day03.txt");

        // Part 1
        TestRunner.RunSolution("Part 1", Part1, testInputPath, realInputPath, expectedTestResult: 357);

        // Part 2
        TestRunner.RunSolution("Part 2", Part2, testInputPath, realInputPath, expectedTestResult: 3121910778619);
    }

    public static long Part1(string input)
    {
        // Input format: Multi-line string of digits
        input = input.Replace("\r", "").Trim();
        var lines = input.Split('\n', StringSplitOptions.RemoveEmptyEntries);

        long totalOutputJoltage = 0;

        foreach (var line in lines)
        {
            var trimmed = line.Trim();
            int maxJoltage = -1;

            // Iterate through all pairs of indices (i, j) such that i < j
            for (int i = 0; i < trimmed.Length; i++)
            {
                int digit_i = trimmed[i] - '0';
                for (int j = i + 1; j < trimmed.Length; j++)
                {
                    int digit_j = trimmed[j] - '0';
                    int joltage = digit_i * 10 + digit_j;
                    if (joltage > maxJoltage)
                    {
                        maxJoltage = joltage;
                    }
                }
            }

            if (maxJoltage != -1)
            {
                totalOutputJoltage += maxJoltage;
            }
        }

        return totalOutputJoltage;
    }

    public static long Part2(string input)
    {
        input = input.Replace("\r", "").Trim();
        var lines = input.Split('\n', StringSplitOptions.RemoveEmptyEntries);

        long totalOutputJoltage = 0;
        int k = 12; // Target length

        foreach (var line in lines)
        {
            var trimmed = line.Trim();
            var stack = new List<int>(k);
            int n = trimmed.Length;

            for (int i = 0; i < n; i++)
            {
                int digit = trimmed[i] - '0';
                int remaining = n - 1 - i;

                // While stack is not empty, current digit is greater than top of stack,
                // and we have enough remaining digits to fill the rest of the sequence
                while (stack.Count > 0 && digit > stack[stack.Count - 1] && stack.Count + remaining >= k)
                {
                    stack.RemoveAt(stack.Count - 1);
                }

                if (stack.Count < k)
                {
                    stack.Add(digit);
                }
            }

            // Construct the number directly from digits
            long maxJoltage = 0;
            for (int i = 0; i < stack.Count; i++)
            {
                maxJoltage = maxJoltage * 10 + stack[i];
            }

            totalOutputJoltage += maxJoltage;
        }

        return totalOutputJoltage;
    }
}

File: AocC/day03.c
#include "day03.h"
#include "runner.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

void day03_run(void) {
    const char* test_path = "../inputs/day03_test.txt";
    const char* real_path = "../inputs/day03.txt";

    run_solution("Part 1", day03_part1, test_path, real_path, 357);
    run_solution("Part 2", day03_part2, test_path, real_path, 3121910778619LL);
}

int64_t day03_part1(const char* input) {
    int64_t total = 0;
    const char* line_start = input;

    while (*line_start) {
        // Find end of line
        const char* line_end = line_start;
        while (*line_end && *line_end != '\n' && *line_end != '\r') {
            line_end++;
        }

        int len = (int)(line_end - line_start);
        if (len > 0) {
            int max_joltage = -1;

            for (int i = 0; i < len; i++) {
                if (!isdigit((unsigned char)line_start[i])) continue;
                int digit_i = line_start[i] - '0';

                for (int j = i + 1; j < len; j++) {
                    if (!isdigit((unsigned char)line_start[j])) continue;
                    int digit_j = line_start[j] - '0';
                    int joltage = digit_i * 10 + digit_j;
                    if (joltage > max_joltage) {
                        max_joltage = joltage;
                    }
                }
            }

            if (max_joltage != -1) {
                total += max_joltage;
            }
        }

        // Move to next line
        while (*line_end == '\n' || *line_end == '\r') {
            line_end++;
        }
        line_start = line_end;
    }

    return total;
}

int64_t day03_part2(const char* input) {
    int64_t total = 0;
    const int k = 12;  // Target length
    const char* line_start = input;

    while (*line_start) {
        // Find end of line
        const char* line_end = line_start;
        while (*line_end && *line_end != '\n' && *line_end != '\r') {
            line_end++;
        }

        int len = (int)(line_end - line_start);
        if (len > 0) {
            // Stack-based monotonic decreasing algorithm
            int stack[20];  // Max k digits
            int stack_size = 0;

            for (int i = 0; i < len; i++) {
                if (!isdigit((unsigned char)line_start[i])) continue;

                int digit = line_start[i] - '0';
                int remaining = 0;

                // Count remaining digits
                for (int j = i + 1; j < len; j++) {
                    if (isdigit((unsigned char)line_start[j])) {
                        remaining++;
                    }
                }

                while (stack_size > 0 && digit > stack[stack_size - 1] &&
                       stack_size + remaining >= k) {
                    stack_size--;
                }

                if (stack_size < k) {
                    stack[stack_size++] = digit;
                }
            }

            // Construct number from stack
            int64_t max_joltage = 0;
            for (int i = 0; i < stack_size; i++) {
                max_joltage = max_joltage * 10 + stack[i];
            }

            total += max_joltage;
        }

        // Move to next line
        while (*line_end == '\n' || *line_end == '\r') {
            line_end++;
        }
        line_start = line_end;
    }

    return total;
}
